File: cpp/127_Word Ladder.cpp
class Solution {
public:
	int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
		unordered_set<string> set(wordList.begin(), wordList.end());
		queue<string> todo;
		todo.push(beginWord);
		int res = 1;
		while (!todo.empty()) {
			int n = todo.size();
			for (int i = 0; i < n; i++) {
				string word = todo.front(); todo.pop();
				if (word == endWord) {
					return res;
				}
				//else erase from set
				//find the next possible word ladder
				//push it into queue
				set.erase(word);
				for (int j = 0; j < size(); ++j) {
					char c = word[j];
					// substitute word[j] with other letters
					for (int k = 0; k < 26; ++k) {
						word[j] = 'a' + k;
						if (set.find(word) != set.end()) {
							//a valid ladder
							todo.push(word);
						}
					}
					word[j] = c;
				}
			}
			res++;	//res is also the level of BFS
		}
		return 0;	//not found
	}
};

// two end BFS
class Solution {
public:
	int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
		unordered_set<string> dict(wordList.begin(), wordList.end());
		unordered_set<string> beginSet, endSet, *set1, *set2;
		if (dict.find(endWord) == dict.end())
			return 0;
		beginSet.insert(beginWord);
		endSet.insert(endWord);

		int res = 2;
		while (!beginSet.empty() && !endSet.empty()) {
			if (beginSet.size() > endSet.size()) {
				swap(beginSet, endSet);
				// make the size of two sets close
				// always explore the smaller set
			}
			unordered_set<string> tmpSet;
			for (auto it = beginSet.begin(); it != beginSet.end(); ++it) {
				string word = *it;
				for (int j = 0; j < word.size(); ++j) {
					char c = word[j];
					for (int k = 0; k < 26; ++k) {
						word[j] = 'a' + k;
						if (endSet.find(word) != endSet.end()) {
							return res;
						}
						if (dict.find(word) != dict.end()) {
							tmpSet.insert(word);
							dict.erase(word);
						}
					}
					word[j] = c;
				}
			}
			swap(beginSet, tmpSet);	//tmpSet become the new beginSet
			res++;
		}
		return 0;	//not found
	}
};

File: java/127.word-ladder.java
import java.util.*;

/*
 * @lc app=leetcode id=127 lang=java
 *
 * [127] Word Ladder
 */

// @lc code=start
class Solution {
    // BFS
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> wordSet = new HashSet<>(wordList);
        if (!wordSet.contains(endWord))
            return 0;
        Queue<String> queue = new LinkedList<>();
        int res = 0;
        queue.offer(beginWord);
        while (!queue.isEmpty()) {
            res++;
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                String curr = queue.poll();
                if (endWord.equals(curr))
                    return res;
                char[] chs = curr.toCharArray();
                for (int j = 0; j < chs.length; j++) {
                    char old = chs[j];
                    for (char c = 'a'; c <= 'z'; c++) {
                        chs[j] = c;
                        if (wordSet.contains(String.valueOf(chs))) {
                            queue.offer(String.valueOf(chs));
                            wordSet.remove(String.valueOf(chs));
                        }
                    }
                    chs[j] = old;
                }
            }
        }
        return 0;
    }

    // two-end BFS
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        if (!wordList.contains(endWord))
            return 0;
        Set<String> beginSet = new HashSet<>(), endSet = new HashSet<>(), wordSet = new HashSet<>(wordList);
        beginSet.add(beginWord);
        endSet.add(endWord);
        int res = 0;
        while (!beginSet.isEmpty() && !endSet.isEmpty()) {
            res++;
            if (beginSet.size() > endSet.size()) {
                // swap. always explore the smaller one
                Set<String> tmp = beginSet;
                beginSet = endSet;
                endSet = tmp;
            }
            Set<String> tmpSet = new HashSet<>();
            for (String curr : beginSet) {
                char[] chs = curr.toCharArray();
                for (int i = 0; i < chs.length; i++) {
                    char old = chs[i];
                    for (char c = 'a'; c <= 'z'; c++) {
                        // substitute one letter in current word
                        chs[i] = c;
                        String word = String.valueOf(chs);
                        if (endSet.contains(word)) // solution found
                            return res + 1;
                        if (wordSet.contains(word)) {
                            tmpSet.add(word); // a valid ladder in this level
                            wordSet.remove(word); // remove visited word
                        }
                    }
                    chs[i] = old;
                }
            }
            beginSet = tmpSet; // tmpSet(the new level of words) becomes the new beginSet
        }
        return 0;
    }

}
// @lc code=end
